### PR TITLE
 fix(consumer): duplicate ID2 breaking DB-mode

### LINKED ISSUE #729: duplicate key-auth credentials cause the ingress controller to fail syncing all future configuration
### Summary

If a KongConsumer references two secrets (they could be different secrets) that contain the same key, then the ingress controller cannot successfully sync any future configuration to Kong.

**Kong Ingress controller version**

0.9.1

**Kong or Kong Enterprise version**

2.0.4

### Kubernetes version

`Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.9", GitCommit:"e09f5c40b55c91f681a46ee17f9bc447eeacee57", GitTreeState:"clean", BuildDate:"2019-05-27T15:58:45Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}`

### Environment

- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): alpine
- **Install tools**: helm
- **Deployment**: DB (postgres)

### What happened

Two secrets with the same key were listed as configuration for a KongConsumer. The ingress controller fails to sync any future configuration until the duplicate key is removed. 

### Expected behavior

The ingress controller should skip duplicate keys or fail safely and allow other unrelated configuration to sync.

### Steps To Reproduce

1. Reference two secrets with the same key in a KongConsumer, or reference the same secret in different KongConsumers.
2. `E0610 20:02:43.392469 1 controller.go:128] unexpected failure updating Kong configuration: inserting key-auth into state: entity already exists`

### PR DESCRIPTION
**Pull request by mflendrich**
fixes #729

Before this PR, KIC failed to apply config to DB-backed Kong, in
case the config contained a Consumer with two credentials having
equivalent ID2 fields (Key for key-auth, ClientID for OAuth2, etc.).

This commit de-duplicates credentials of each type, by replacing the
Consumer store for credentials with a map (ID2 -> credential object),
and updates relevant tests.

Also, this commit implements consistent credential ordering.

---

refactor(controller): implement sortByString

This REVERSES the sorting order from reverse to normal.

The history behind reverse order is purely accidental. There are no
known reasons in favor of one or the other. What matters is the order
being deterministic.

---

test(controller): toDeckContent sorts creds

---

test(kongstate): deduplicates credentials

### CODE REVIEW

**Comment by rainest** [comment]
How does this handle sort order for credentials that have the same ID2? It's effectively random, correct?

I think this introduces a footgun that allows you to inadvertently replace in-use configuration, e.g.

1. UserA creates a JWT credential with key `secret` and RSApubkeyA. Their application starts using this credential.
2. UserB, having not consulted with UserA, creates another JWT credential with key `secret` and RSApubkeyB.
3. The controller arbitrarily chooses UserB's credential after sorting, rejects UserA's credential as a duplicate, and updates state to include UserB's credential only.
4. UserA's application starts receiving 401s.

This is somewhat mitigated because this only handles credentials for a single consumer, not credentials on separate consumers (we'll still attempt to apply those and hit the same admin API error). We'd need to further tiebreak by credential creation time to avoid that.

File: internal/ingress/controller/kong.go
Comment (shaneutt) [inline internal/ingress/controller/kong.go:295]: golint: public symbol should have a godoc comment

File: internal/ingress/controller/kong.go
Comment (shaneutt) [inline internal/ingress/controller/kong.go:416]: > refactor(controller): implement sortByString
> This REVERSES the sorting order from reverse to normal.
> The history behind reverse order is purely accidental. There are no
> known reasons in favor of one or the other. What matters is the order
> being deterministic.

Could you help me to better understand the impetus for this change, just for my own edification? If there's no known reasons in favor of one versus the other, is this change intended simply to make sure our ordering is intentional going forward for consistency? Or is there further context at play here?

Comment (shaneutt) [inline internal/ingress/controller/kong.go:416]: on the off chance we put this here and then forget about it, but then someone else wants to implement the same logic somewhere else later would it be worthwhile to move this into `internal/*/utils` or something like that do you think?

File: internal/ingress/controller/kong_test.go
Comment (shaneutt) [inline internal/ingress/controller/kong_test.go:12]: yay I love testify ^_^

File: internal/ingress/controller/kong_test.go
Comment (shaneutt) [inline internal/ingress/controller/kong_test.go:132]:
```suggestion
			name: "sorts credentials consistently",
```

File: internal/ingress/controller/parser/kongstate/consumer.go
Comment (shaneutt) [inline internal/ingress/controller/parser/kongstate/consumer.go:31]: question: I'm simply curious for my own learning what instigated this change? I would assume this is for convenience to avoid having to initialize these manually every time you make a new `Consumer`, is that pretty much it?

And that being the case, we should perhaps add some documentation for the `NewConsumer()` public function which will point out this behavior for readers of our Godoc.

File: internal/ingress/controller/parser/kongstate/consumer.go
Comment (shaneutt) [inline internal/ingress/controller/parser/kongstate/consumer.go:27]: golint: public symbol needs godoc comments

**Review by shaneutt — COMMENTED** [review]
Generally LGTM, just a couple of small comments :+1:

File: internal/ingress/controller/parser/kongstate/consumer.go
Comment (mflendrich) [reply internal/ingress/controller/parser/kongstate/consumer.go:31]: Before this PR, a zero-initialized `Consumer` was a valid consumer. Now it is not because of introduction of `map`s. Hence `NewConsumer()`. `NewConsumer` could simply initialize all fields of `Consumer`, but that would be tedious for tests. In tests we often need "a minimal consumer with one specific map initialized, and the rest left empty".

This is an idiom I've been following in such scenario, but for my own learning I'd love to see how others deal with this problem (of partial initialization for testing purposes).

File: internal/ingress/controller/kong.go
Comment (mflendrich) [reply internal/ingress/controller/kong.go:416]: `toDeckContent()` generates a big struct that gets eventually fed into `Kong`. #404 added an optimization: after that PR, KIC refrains from feeding a new config to Kong if that config's hash hasn't changed since last known version.

In order to ensure that equivalent configs hash in a predictable manner, it's important to guarantee that the order of elements in slices (otherwise irrelevant) remains deterministic. Hence sorting.

The order of sorted elements doesn't matter though (as long as it's deterministic). #404 happened to implement sorting in reverse lexicographic order. I found the "reverse" bit unnecessarily confusing. Therefore,

> is this change intended simply to make sure our ordering is intentional going forward for consistency

is an accurate depiction of the situation.

Comment (mflendrich) [reply internal/ingress/controller/kong.go:416]: 1. Historically, exporting util functions followed direct intent to reuse, not the other way around. Not saying that it's the only right way - just we've happened to work like that so far. This may be my own bias, but a function being exported or not doesn't affect my ability to remember its existence when I need a similar one elsewhere.
2. This function is a 2-liner. Exporting and importing proactively feels like an overkill to me for this degree of complexity.
3. An exported function would need individual tests. Here, this helper function is tested as a part of a bigger thing.

**Review by shaneutt — DISMISSED** [review]
Good explanations in the code review, thank you.

LGTM :heavy_check_mark:

**Comment by mflendrich** [comment]
> How does this handle sort order for credentials that have the same ID2? It's effectively random, correct?

The order is directly as returned from informer stores. Therefore,  we need to treat it as effectively not guaranteed to be anything meaningful.

> I think this introduces a footgun that allows you to inadvertently replace in-use configuration, e.g.
> 
>     1. UserA creates a JWT credential with key `secret` and RSApubkeyA. Their application starts using this credential.
> 
>     2. UserB, having not consulted with UserA, creates another JWT credential with key `secret` and RSApubkeyB.
> 
>     3. The controller arbitrarily chooses UserB's credential after sorting, rejects UserA's credential as a duplicate, and updates state to include UserB's credential only.
> 
>     4. UserA's application starts receiving 401s.

That is correct, and this is a part of a bigger problem: given two conflicting entities that made it through our admission validations, KIC needs to distill _some_ valid config from it and apply it to Kong. Otherwise the only option is for KIC's sync loop to lock up, which is terrible user experience.

The problem is that a fair "tie breaker" for conflicting configs is an unsolved problem. The best we can do today is to drop an arbitrary subset that makes the config no longer nonsensical.

> This is somewhat mitigated because this only handles credentials for a single consumer, not credentials on separate consumers (we'll still attempt to apply those and hit the same admin API error). We'd need to further tiebreak by credential creation time to avoid that.

I would like to avoid using creation time because it introduces an external factor (creation time) to KIC's semantics. This external factor is likely to surprisingly change if the user, for example, restores their k8s objects from a backup, or from a gitops repository. Therefore I don't see a clear benefit of using creation time rather than picking an arbitrary resource. I'm all in for deterministic and well-defined conflict resolution, but I would rather see something that relies on the configuration that the user controls (e.g., in case of conflicting resources, drop one that has a lexicographically greater namespace+name).

I recommend leaving this as-is and creating an issue to be solved later (deterministic conflict resolution based on user-controlled factors). WDYT?

**Comment by mflendrich** [comment]
As agreed in a zoom call (@rainest @mflendrich @shaneutt), the choice of preferred solution to this problem (hashing the entire object vs hashing a set of fields) depends on whether components depending on these IDs will aggregate data within a high-level business object that potentially maps to multiple IDs.

**Comment by mflendrich** [comment]
Solution blocked by discrepancy between DB-enabled and DB-less mode of Kong (as described in https://github.com/Kong/kong/issues/6732). Semantics of conflict resolution to be defined following agreement with Kong/kong maintainers on the desired semantics of duplicate entries.

**Comment by rainest** [comment]
Re discussion elsewhere on whether to use creation timestamp or something else, I'd like to step back and reject the underlying premise:

> This is a part of a bigger problem: given two conflicting entities that made it through our admission validations, KIC needs to distill some valid config from it and apply it to Kong.

This is hamstrung by a lack of any valid configuration in this scenario. If you've created two credentials that collide, we can only apply one, and which we choose to apply affects the proxy behavior. I originally suggested creation timestamp because it seemed least likely to result in a change in proxy behavior, since the older resource would have a greater chance of being applied first, but it's not a great solution either.

Any tiebreaker we choose is effectively arbitrary, but user-controlled tiebreakers pose an odd question: if you know that two resources collide, why would you set the tiebreaker such that one resource is used rather than either deleting one of the resources or choosing non-colliding values? If you don't know, you could try at random until your configuration wins, but I don't think it's something we'd want to encourage.

> Otherwise the only option is for KIC's sync loop to lock up, which is terrible user experience.

Agreed, but it at least fails relatively safely, since your proxy will continue using existing configuration. Assuming you're not aware of the conflict (again, if you are, you should resolve it by other means), using a tiebreaker results in one of the following:
- New configuration loses the tiebreaker. The controller doesn't actually apply your configuration, and reports it applied configuration successfully. You don't know why your configuration isn't working.
- New configuration wins the tiebreaker. The controller reports success, and as far as you're concerned, everything is working great. However, someone else's configuration has spontaneously stopped working, and they probably aren't happy. They don't know why this happened, because the controller just reports a successful sync, and their configuration wasn't deleted.

These are both bad user experience, and potentially more disruptive, since the latter obliterates in-use configuration.

The real issue with the sync loop blocking is that we cannot discard configuration piecemeal: it'd be a much smaller problem if it did not lock up entirely and prevent sync of other (valid) configuration added after. This is hard to solve, but I don't think attempting to mask collisions is a good way to handle it. Absent some way to report (and ideally reject) the problem configuration, I think the only safe option is to let the status quo (sync block) stand.

**Comment by mflendrich** [comment]
> Any tiebreaker we choose is effectively arbitrary, but user-controlled tiebreakers pose an odd question: if you know that two resources collide, why would you set the tiebreaker such that one resource is used rather than either deleting one of the resources or choosing non-colliding values? If you don't know, you could try at random until your configuration wins, but I don't think it's something we'd want to encourage.

I fully agree that we don't want to encourage KIC with invalid configurations.

That said, as a matter of fact, users will keep feeding invalid configurations to KIC. What can we do about it? I see three layers of defense here:

1. prevent invalid configurations from being fed to KIC altogether
2. given invalid configuration, warn the user visibly
3. given invalid configuration, fail gracefully

(1) is achieved by CRD schemas and webhook validations. None of these is capable of validating unique constraints across all resources in the cluster, other than resource name. There are measures we can theoretically take to prevent _some_ of these cases, but conflicts will fall through the cracks. Outside of the scope of this issue I think, but we can agree here to implement such a solution separately to improve UX.

(2) KIC does this today by print warning log entries. A better way to do that would be in `status` fields of affected resources. A perfect way would also to enhance KIC with a Kubernetes resource for generated configuration (likely happening as part of #702 - @shaneutt) and put the overall success/failure/errors/warnings in the `status` thereof.

(3) is the clue here.

It's quite clear that if KIC config contains 2 or more conflicting entities, there's no theoretical possibility to apply a working and _correct_ config for that subset of configuration because there's no "correct" way to do it. This is obviously an unrecoverable failure scenario.

In an unrecoverable failure scenario one thing that we have under control is the _scope_ of that failure. And:
- if KIC stops syncing config altogether, then that's equivalent to the entire KIC failing. The user loses the ability to configure any other workloads and consumers, possibly causing an outage elsewhere. Imagine a multi-tenant k8s cluster with one ingress controller where:
    - tenant A accidentally creates a route on a critical workload that causes a disaster (e.g. directs all traffic to a backend service that should be cached instead)
    - tenant B creates a consumer conflict in their toy app
    - tenant A reverts the route change (ineffectively) and sees their application melt down further
- if KIC contains the failure within conflicting resources, operations for other resources continue without downgrade

So, the question here is whether, in case of failure, we choose to degrade only to the extent of the conflicted resources, or to degrade everything.

Here are the concerns about the proposed solution I've understood from https://github.com/Kong/kubernetes-ingress-controller/pull/1007#issuecomment-769452064 (please correct me if I missed something):

1. KIC syncing new configuration after a conflicts was introduced can cause the user to oversee the conflict
2. adding new conflicting config may obliterate the old conflicting config
3. we cannot discard configuration piecemeal

On (1): I believe that we can solve this by providing better feedback about configuration errors - using `status` fields on (existing and future) resoures. As soon as we have a resource for the entire rendered configuration, its `status` would be the most natural place to post such errors - as well as `status` of conflicted resources themselves.

On (2): the premise that old config is more meaningful than new config, and should always win, incompatible with the Kubernetes configuration model that is based on eventual consistency and no temporal relationships between entities.
Furthermore, I view the solution of preserving old configuration by locking up the entire sync loop as throwing the baby out with the bathwater. Alas, I don't see any better way to achieve preserving old configuration.

On (3): why not? This is a serious question. Maybe I'm missing some hidden complexity. My thinking is: "walk through all the resources in a deterministic, well known order. If the currently processed resource is conflicted with the already seen state, warn loudly and assume the fiction that it didn't exist in the config".

Therefore I propose that:
- we do improve user feedback by utilizing status fields to expose conflicts,
- we do minimize the configuration impact of a configuration error by leaving everything but the conflicted resources unaffected,
- we drop the idea to break ties by entity age (unless we find a way that doesn't use timestamps and doesn't lock up too many resources)

WDYT @rainest?

**Comment by mflendrich** [comment]
Removed the "blocked" label because the blocking premise (need to rely on Kong/kong maintainers' input about the discrepancy between db and dbless) is obsolete.

**Comment by mflendrich** [comment]
Closing this PR - migration of KIC to Railgun will command a different solution.

**Pull request closed without merging**
